File: 40.组合总和-ii.go
/*
 * @lc app=leetcode.cn id=40 lang=golang
 *
 * [40] 组合总和 II
 */
package main

import "sort"

// @lc code=start
func combinationSum2(candidates []int, target int) [][]int {
	n := len(candidates)
	res := [][]int{}
	if n == 0 {
		return res
	}
	sort.Ints(candidates)
	path := []int{}
	res = dfs2(candidates, 0, n, target, path, res)
	return res
}

func dfs2(candidates []int, begin, n, target int, path []int, res [][]int) [][]int {
	if target == 0 {
		res = append(res, append([]int(nil), path...))
		return res
	}
	for i := begin; i < n; i++ {
		if target-candidates[i] < 0 {
			break
		}
		if i > begin && candidates[i] == candidates[i-1] {
			continue
		}
		path = append(path, candidates[i])
		res = dfs2(candidates, i+1, n, target-candidates[i], path, res)
		path = path[:len(path)-1]
	}
	return res
}

// @lc code=end

File: 40.组合总和-ii.cs
/*
 * @lc app=leetcode.cn id=40 lang=csharp
 *
 * [40] 组合总和 II
 */

// @lc code=start
public class Solution
{
    public IList<IList<int>> CombinationSum2(int[] candidates, int target)
    {
        int len = candidates.Length;
        IList<IList<int>> res = new List<IList<int>>();
        if (len == 0)
        {
            return res;
        }
        Array.Sort(candidates);
        Stack<int> path = new Stack<int>();
        Dsf2(candidates, 0, len, target, path, res);
        return res;
    }

    public void Dsf2(int[] candidates, int begin, int len, int target, Stack<int> path, IList<IList<int>> res)
    {
        if (target == 0)
        {
            List<int> newPath = new(path);
            newPath.Sort();
            res.Add(newPath);
            return;
        }
        for (int i = begin; i < len; i++)
        {
            if (target - candidates[i] < 0)
            {
                break;
            }
            if (i > begin && candidates[i] == candidates[i - 1])
            {
                continue;
            }
            path.Push(candidates[i]);
            Dsf2(candidates, i + 1, len, target - candidates[i], path, res);
            path.Pop();
        }
    }
}
// @lc code=end

